tests/test_chiral_restraints.py

```diff
@@ -626,5 +636,64 @@ def assert_same_signs(a, b):
         frame_bond_vols = get_chiral_bond_volumes(f)
         # f = f - np.mean(f, axis=0)
         # writer.write_frame(f*10)
-        assert_same_signs(frame_atom_vols, ref_chiral_atom_vols)
-        assert_same_signs(frame_bond_vols, ref_chiral_bond_vols)
+        if check_chiral_atoms:
+            assert_same_signs(frame_atom_vols, ref_chiral_atom_vols)
+        if check_chiral_bonds:
+            assert_same_signs(frame_bond_vols, ref_chiral_bond_vols)
+
+
+def test_chiral_inversion_in_single_topology():
+    very_short_hrex_params = replace(
+        DEFAULT_HREX_PARAMS, n_frames=2, n_eq_steps=2, steps_per_frame=2, hrex_params=HREXParams(n_frames_bisection=2)
+    )
+    ff = Forcefield.load_default()
+
+    # smi_a simplified from
+    # https://github.com/rdkit/rdkit/blob/06b135b4d54ea0c34d8a488df730390267bf5a36/Code/GraphMol/testChirality.cpp#L3213C18-L3213C41
+    smi_a = "C[C@]1(F)C[C@]([H])(O)C1"
+    smi_b = "C[C@]1(F)C[C@@]([H])(O)C1"  # invert one center, relative to mol_a
+    mol_a = Chem.AddHs(Chem.MolFromSmiles(smi_a))
+    mol_b = Chem.AddHs(Chem.MolFromSmiles(smi_b))
+    mol_a.SetProp("_Name", "A")
+    mol_b.SetProp("_Name", "B")
+    AllChem.EmbedMolecule(mol_a, randomSeed=2024)
+
+    def generate_more_complete_mapping():
+        """map whole ring + one substituent"""
+        substruct = Chem.MolFromSmiles("C1CCC1C")
+        atom_map = dict(zip(mol_a.GetSubstructMatch(substruct), mol_b.GetSubstructMatch(substruct)))
+        conf_a = mol_a.GetConformer()
+        coord_map = {atom_map[idx]: conf_a.GetAtomPosition(idx) for idx in atom_map}
+        AllChem.EmbedMolecule(
+            mol_b,
+            randomSeed=2024,
+            coordMap=coord_map,
+            useRandomCoords=True,
+            clearConfs=True,
+            enforceChirality=True,
+            useBasicKnowledge=True,
+        )
+        core = get_cores(mol_a, mol_b, **DEFAULT_ATOM_MAPPING_KWARGS)[0]
+        assert len(core) == 15
+        return mol_a, mol_b, core
+
+    def generate_less_complete_mapping():
+        """map part of ring + other substituent"""
+        substruct = Chem.MolFromSmiles("C1CCC1O")
+        atom_map = dict(zip(mol_a.GetSubstructMatch(substruct), mol_b.GetSubstructMatch(substruct)))
+        conf_a = mol_a.GetConformer()
+        coord_map = {atom_map[idx]: conf_a.GetAtomPosition(idx) for idx in atom_map}
+        AllChem.EmbedMolecule(mol_b, randomSeed=2045, coordMap=coord_map)
+        core = get_cores(mol_a, mol_b, **DEFAULT_ATOM_MAPPING_KWARGS)[0]
+        assert len(core) == 10
+
+        return mol_a, mol_b, core
+
+    # check that it runs
```

Comment:
Would be it useful to generate an explicit ring-breaking atom-mapping as well (replicating the original observation)?

Reply:
Will make this more explicit.

(Subtlety: original observation was _not_ a ring-breaking transformation, but a partially mapped ring due to incomplete alignment, which I tried to emulate in generating these two different alignments / atom-mappings.)

```python
    def generate_more_complete_mapping():
        """map whole ring + one substituent"""
```
produces something like:

<img width="708" alt="image" src="https://github.com/proteneer/timemachine/assets/5759036/0fc5fb2e-25a6-4ce5-b6ec-0bc396c4ad35">


and 
```python
    def generate_less_complete_mapping():
        """map part of ring + other substituent"""
```
produces something like
<img width="708" alt="image" src="https://github.com/proteneer/timemachine/assets/5759036/b278b873-8c22-4efb-bed0-953bbe2bb3f7">

Reply:
Ah got it, thank you for including the pictures, this resolves my concern. I was looking at the line `substruct = Chem.MolFromSmiles("C1CCC1O")` and thinking that it also implied the core being a complete ring somehow (since the ring atom positions of `mol_b` are then fixed using the `coordMap`) and are then fully mapped during the MCS. But I missed that `get_cores` will always end up returning a core with a partial ring due to the use of chiral filters during the MCS, even if all distance checks pass.

Reply:
Sorry, 2 independent confusing things, one possibly important, one inconsequential:

> get_cores will always end up returning a core with a partial ring due to the use of chiral filters during the MCS, even if all distance checks pass.

For future reference, this is not a correct description of the behavior of the MCS implementation.

The demapped ring atom (atom idx 1) differed by only 1.03 A between mol_a and mol_b (which is indeed below [ring_cutoff 1.2 A](https://github.com/proteneer/timemachine/blob/60b37e6fc0e541c8129575e585702d9cd1d3aa6a/timemachine/constants.py#L29)) . But increasing the distance cutoffs even further **does** cause this atom to be mapped:

<img width="660" alt="image" src="https://github.com/proteneer/timemachine/assets/5759036/39922da8-c500-4009-bb87-256109fd3222">

(The partial demapping of this ring **does** appear to depend on the distance checks, and get_cores can return a complete ring match when a sufficient number of additional distance checks pass. But it may also interact with chiral filters or other filters -- I do not have a good understanding of what guarantees might hold for the overall discrete search routine including all filters.)

> was looking at the line substruct = Chem.MolFromSmiles("C1CCC1O") and thinking that it also implied the core being a complete ring somehow (since the ring atom positions of mol_b are then fixed using the coordMap) and are then fully mapped during the MCS.

(Detail that's not very pertinent: `EmbedMolecule(..., coord_map)` almost never succeeds in exactly snapping the selected substructures to requested coordinates with default settings. But toggling `useRandomCoords=True` [usually does the trick](https://greglandrum.github.io/rdkit-blog/posts/2023-02-10-more-on-constrained-embedding.html). (Note that the atoms in the matched substructure are exactly aligned in the atom map produced by `generate_more_complete_mapping()` where `useRandomCoords=True`, but not in the atom map produced by `generate_less_complete_mapping()` where `useRandomCoords=False` by default.) This was an arbitrary and implicit way to generate more or less `mol_a`-aligned conformers for `mol_b` -- removed from test script and moved outputs into tests/data/*.sdf in https://github.com/proteneer/timemachine/pull/1243/commits/79bb03df0571a42d3e292a5e098db047bc83f246 )

Reply:
(brief pedagogical note after offline discussion with @maxentile: for the last row, flipping the molecule allows for a full mapping, if every atom is allowed to map to every atom.)